url4c41254592_topskychen_AuthTopk_tgz-pJ8-quickhull3d_QuickHull3DTestJ8/src/spatialindex/core/Region.java

```diff
@@ -0,0 +1,433 @@
+
+	//
+	// Cloneable interface
+	//
+
+	@Override
+	public Object clone() {
+		return new Region(m_pLow, m_pHigh);
+	}
+
+	//
+	// IShape interface
+	//
+
+	@Override
+	public boolean intersects(final IShape s) {
+		if (s instanceof Region) {
+			return intersects((Region) s);
+		}
+
+		if (s instanceof Point) {
+			return contains((Point) s);
+		}
+
+		throw new IllegalStateException("intersects: Not implemented yet!");
+	}
+
+	@Override
+	public boolean contains(final IShape s) {
+		if (s instanceof Region) {
+			return contains((Region) s);
+		}
+
+		if (s instanceof Point) {
+			return contains((Point) s);
+		}
+
+		throw new IllegalStateException("contains: Not implemented yet!");
+	}
+
+	@Override
+	public boolean touches(final IShape s) {
+		if (s instanceof Region) {
+			return touches((Region) s);
+		}
+
+		if (s instanceof Point) {
+			return touches((Point) s);
+		}
+
+		throw new IllegalStateException("touches: Not implemented yet!");
+	}
+
+	@Override
+	public double[] getCenter() {
+		double[] pCoords = new double[m_pLow.length];
+
+		for (int cIndex = 0; cIndex < m_pLow.length; cIndex++) {
+			pCoords[cIndex] = (m_pLow[cIndex] + m_pHigh[cIndex]) / 2.0;
+		}
+
+		return pCoords;
+	}
+
+	@Override
+	public long getDimension() {
+		return m_pLow.length;
+	}
+
+	@Override
+	public Region getMBR() {
+		return new Region(m_pLow, m_pHigh);
+	}
+
+	@Override
+	public double getArea() {
+		double area = 1.0;
+
+		for (int cIndex = 0; cIndex < m_pLow.length; cIndex++) {
+			area *= m_pHigh[cIndex] - m_pLow[cIndex];
+		}
+
+		return area;
+	}
+
+	@Override
+	public double getMinimumDistance(final IShape s) {
+		if (s instanceof Region) {
+			return getMinimumDistance((Region) s);
+		}
+
+		if (s instanceof Point) {
+			return getMinimumDistance((Point) s);
+		}
+
+		throw new IllegalStateException("getMinimumDistance: Not implemented yet!");
+	}
+
+	/**
+	 * return relation code with another region, which code is described in
+	 * IVisitor
+	 * 
+	 * @param r
+	 * @return
+	 */
+	public int getRelationCode(IShape r) {
+		if (this.intersects(r))
+			return IVisitor.TYPE_INTERSECT;
+		else if (this.contains(r))
+			return IVisitor.TYPE_INSIDE;
+		else
+			return IVisitor.TYPE_OUTSIDE;
+	}
+
+	public boolean intersects(final Region r) {
+		if (m_pLow.length != r.m_pLow.length) {
+			throw new IllegalArgumentException("intersects: Shape has the wrong number of dimensions.");
+		}
+
+		for (int cIndex = 0; cIndex < m_pLow.length; cIndex++) {
+			if (m_pLow[cIndex] > r.m_pHigh[cIndex] || m_pHigh[cIndex] < r.m_pLow[cIndex]) {
+				return false;
+			}
+		}
+		return true;
+	}
+
+	public boolean contains(final Region r) {
+		if (m_pLow.length != r.m_pLow.length) {
+			throw new IllegalArgumentException("contains: Shape has the wrong number of dimensions.");
+		}
+
+		for (int cIndex = 0; cIndex < m_pLow.length; cIndex++) {
+			if (m_pLow[cIndex] > r.m_pLow[cIndex] || m_pHigh[cIndex] < r.m_pHigh[cIndex]) {
+				return false;
+			}
+		}
+		return true;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 222:231, 271:280.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.